modules/light-clients/08-wasm/light_client_module.go

```diff
@@ -276,14 +277,69 @@ func (lcm LightClientModule) RecoverClient(ctx sdk.Context, clientID, substitute
 	return clientState.CheckSubstituteAndUpdateState(ctx, cdc, clientStore, substituteClientStore, substituteClient)
 }
 
-// // Upgrade functions
-// // NOTE: proof heights are not included as upgrade to a new revision is expected to pass only on the last
-// // height committed by the current revision. Clients are responsible for ensuring that the planned last
-// // height of the current revision is somehow encoded in the proof verification process.
-// // This is to ensure that no premature upgrades occur, since upgrade plans committed to by the counterparty
-// // may be cancelled or modified before the last planned height.
-// // If the upgrade is verified, the upgraded client and consensus states must be set in the client store.
-// // DEPRECATED: will be removed as performs internal functionality
-func (LightClientModule) VerifyUpgradeAndUpdateState(ctx sdk.Context, clientID string, newClient []byte, newConsState []byte, upgradeClientProof, upgradeConsensusStateProof []byte) error {
-	return nil
+// Upgrade functions
+// NOTE: proof heights are not included as upgrade to a new revision is expected to pass only on the last
+// height committed by the current revision. Clients are responsible for ensuring that the planned last
+// height of the current revision is somehow encoded in the proof verification process.
+// This is to ensure that no premature upgrades occur, since upgrade plans committed to by the counterparty
+// may be cancelled or modified before the last planned height.
+// If the upgrade is verified, the upgraded client and consensus states must be set in the client store.
+func (lcm LightClientModule) VerifyUpgradeAndUpdateState(
+	ctx sdk.Context,
+	clientID string,
+	newClient []byte,
+	newConsState []byte,
+	upgradeClientProof,
+	upgradeConsensusStateProof []byte,
+) error {
+	clientType, _, err := clienttypes.ParseClientIdentifier(clientID)
+	if err != nil {
+		return err
+	}
+
+	if clientType != types.Wasm {
+		return errorsmod.Wrapf(clienttypes.ErrInvalidClientType, "expected: %s, got: %s", types.Wasm, clientType)
+	}
+
+	var (
+		cdc               = lcm.keeper.Codec()
+		newClientState    exported.ClientState
+		newConsensusState exported.ConsensusState
+	)
+
+	if err := cdc.UnmarshalInterface(newClient, &newClientState); err != nil {
```

Comment:
I wonder if we could unmarshal directly into the concrete type and drop the cast assertions that follow?

Reply:
As discussed on Monday, we need to unmarshal into the interface.